src/solvers/smt2/smt2_conv.cpp

```diff
@@ -4367,27 +4367,23 @@ void smt2_convt::find_symbols(const exprt &expr)
       out << ")" << "\n";
     }
   }
-  else if(expr.id()==ID_array_of)
+  else if(expr.id() == ID_array_of)
   {
-    if(defined_expressions.find(expr)==defined_expressions.end())
+    if(defined_expressions.find(expr) == defined_expressions.end())
     {
+      const array_of_exprt &array_of_expr = to_array_of_expr(expr);
+
       const irep_idt id =
         "array_of." + std::to_string(defined_expressions.size());
-      out << "; the following is a substitute for lambda i. x" << "\n";
-      out << "(declare-fun " << id << " () ";
-      convert_type(expr.type());
-      out << ")" << "\n";
-
-      // use a quantifier instead of the lambda
-      #if 0 // not really working in any solver yet!
-      out << "(assert (forall ((i ";
-      convert_type(array_index_type());
-      out << ")) (= (select " << id << " i) ";
-      convert_expr(expr.op0());
-      out << ")))" << "\n";
-      #endif
+      out << "(define-fun " << id << " () ";
+      convert_type(array_of_expr.type());
+      out << " ((as const ";
```

Comment:
I'm pretty sure `as const` isn't in SMT-LIB.

Reply:
I am not sure about the status of `as const` in SMTLIB2 standard, but both Z3 and CVC4 accept `as const`, and we anyway didn't have a working solution for this.

Reply:
I will make this translation solver-specific then -- only emit SMT2 output if `solver` is either Z3 or CVC4.

Reply:
Yes, it is fine to use extensions but they must not be in the default.  Z3 and CVC4 are not the only solvers, they are not even the only solvers that people use with CBMC.

Reply:
I agree. Please take another look and see if the changes look okay. I introduced the `use_as_const` field to track solvers that support `as const`.